yt_dlp/extractor/instagram.py

```diff
@@ -358,12 +356,29 @@ def _extract_embed_url(webpage):
 
     def _real_extract(self, url):
         video_id, url = self._match_valid_url(url).group('id', 'url')
-        webpage, urlh = self._download_webpage_handle(url, video_id)
+        info = self._download_json(
+            f'https://i.instagram.com/api/v1/media/{_id_to_pk(video_id)}/info/', video_id, fatal=False,
+            headers={
+                'Accept': '*',
+                'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36',
+                'Authority': 'www.instagram.com',
+                'Referer': 'https://www.instagram.com',
+                'x-ig-app-id': '936619743392459',
+            })
+        if info:
+            return self._extract_product(info['items'][0])
+
+        self.report_warning('The API is locked behind login. '
+                            'Trying with webpage (Some metadata might be missing)', video_id)
+        webpage, urlh = self._download_webpage_handle(url, video_id, errnote='Requested content was not found')
```

Comment:
Despite this message, it seems to me that the API is actually missing metadata as compared to the webpage. In that case, the order should be reversed

Reply:
i guess what you have done is preety much sufficient

Reply:
?

Reply:
I have not modified the order. I only cleaned it up to follow our conventions and didn't change any logic

Reply:
Then the existing method is ok, i guess no changes are needed

Reply:
Actually I can't understand what you are trying to point out. Also please mention what needs to be changed

Reply:
> Despite this message, it seems to me that the API is actually missing metadata as compared to the webpage.

Am I wrong about this? If not, the webpage-based extraction has to be attempted first. It should fallback to API only if that fails

Reply:
was this explanation sufficient or i left something out

Reply:
now the webpage extraction happens first then API but if api fails it uses data from embed extraction. Is that ok?

Reply:
huh? I say webpage extraction is superior and you completely remove it?

There are 3 steps. Webpage, API, embed. Of these, webpage gives most metadata and should be attempted first. If it fails, use API. Use embed page only if both fail!

Reply:
Webpage extraction is useless as now it no longer contains any usefull data. So as you said then it must be in the order API, embed.

Reply:
The webpage is only useless ONLY if you are logged in. Your code breaks when when not logged in

Reply:
Yea but since the webpage hold nothing and makes a graphql call for metadata so the current order is graphql, API, embed.